rust/lance-index/src/scalar/bloomfilter.rs

```diff
@@ -0,0 +1,1769 @@
+            })?
+            .as_any()
+            .downcast_ref::<arrow_array::UInt64Array>()
+            .ok_or_else(|| {
+                Error::invalid_input(
+                    "BloomFilterIndex: 'block_size' column is not UInt64",
+                    location!(),
+                )
+            })?;
+
+        let bloom_filter_data_col = data
+            .column_by_name("bloom_filter_data")
+            .ok_or_else(|| {
+                Error::invalid_input(
+                    "BloomFilterIndex: missing 'bloom_filter_data' column",
+                    location!(),
+                )
+            })?
+            .as_any()
+            .downcast_ref::<arrow_array::BinaryArray>()
+            .ok_or_else(|| {
+                Error::invalid_input(
+                    "BloomFilterIndex: 'bloom_filter_data' column is not Binary",
+                    location!(),
+                )
+            })?;
+
+        let num_blocks = data.num_rows();
+        let mut blocks = Vec::with_capacity(num_blocks);
+
+        for i in 0..num_blocks {
+            let bloom_filter_data = if bloom_filter_data_col.is_valid(i) {
+                bloom_filter_data_col.value(i).to_vec()
+            } else {
+                Vec::new()
+            };
+
+            blocks.push(BloomFilterStatistics {
+                fragment_id: fragment_id_col.value(i),
+                block_start: block_start_col.value(i),
+                block_size: block_size_col.value(i) as usize,
+                bloom_filter_data,
+            });
+        }
+
+        Ok(Self {
+            blocks,
+            data_type,
+            number_of_items,
+            number_of_bytes,
+            store,
+            fri,
+            index_cache,
+        })
+    }
+
+    fn evaluate_block_against_query(
+        &self,
+        block: &BloomFilterStatistics,
+        query: &SargableQuery,
+    ) -> Result<bool> {
+        // If no bloom filter data is available, be conservative and return true
+        if block.bloom_filter_data.is_empty() {
+            return Ok(true);
+        }
+
+        // Try to reconstruct the bloom filter from the stored data
+        let sbbf = match Sbbf::new(&block.bloom_filter_data) {
+            Ok(sbbf) => sbbf,
+            Err(_) => {
+                // If we can't reconstruct the bloom filter, be conservative
+                return Ok(true);
+            }
+        };
+
+        match query {
+            // TODO: Shall we care about null here?
+            SargableQuery::IsNull() => {
+                // For null values, we could use a special marker in the bloom filter
+                // For now, conservatively assume blocks might contain nulls
+                Ok(true)
+            }
+            SargableQuery::Equals(target) => {
+                if target.is_null() {
+                    // Handle null values - conservatively return true for now
```

Comment:
I think we can just add a `has_null` flag in the metadata to make it work

Reply:
let me put more thoughts into it